### ISSUE TITLE
`containsExactly` does not work properly with maps not using equals to compare keys

### ISSUE THREAD

**Issue by joel-costigliola** [assertj/assertj#2165]
#### Summary

This test should pass but fails.

```java
  @Test
  void should_pass_if_case_insensitive_actual_contains_given_entries_in_rder() {
    // GIVEN
    actual = new CaseInsensitiveMap<>();
    actual.put( "Color", "green" );
    actual.put( "NAME", "Yoda" );

    // THEN
    maps.assertContainsExactly(someInfo(), actual, entry("COLOR", "green"), entry("Name", "Yoda"));
  }
```

The underlying issue is when `containsExactly` verifies that the entries are in the same order, it ends up comparing `"Color"` against `"COLOR"` with `equals` instead of the actual map key comparison strategy (case insensitive string in the example).

It's not yet clear if there is a way to address this issue, if not we would have to mention the limitation in the javadoc.

**Comment by scordio** [comment]
This issue can be fixed in a similar way to #2167.

However, the tests might be a bit more complex due to the additional failure in case `actual` contains the given entries but in a different order.

**Comment by wouterpolet** [comment]
Hi! I think I can fix this bug.
I'll create a PR for it if that's okay

**Comment by scordio** [comment]
Thanks for your help, @wouterpolet!

This issue might be not trivial so that's not the best candidate for a first contribution, but it's totally fine and highly appreciated if you want to look at it 🙂

As I mentioned in my previous comment, we expect the solution to be similar to what was done in #2167. However, that implementation evolved and there might be more to be considered (e.g., #2379) so best to look at the current state of `containsOnly` assertions.

One of the main challenges here is probably the testing pattern: we would like to have something similar to `Maps_assertContainsOnlyKeys_Test` for thorough coverage, going through all the map implementations available in our test classpath.

**Comment by scordio** [comment]
The related change has been reverted in #3321 and the topic should be evaluated again.

**Comment by msmuenchen** [comment]
I'm hitting this problem with a `LinkedList<String, byte[]>`.

```java
assertThat(ChannelSecret.lookupChannelSecret(hexFormat.parseHex("3c"))).contains(Map.entry("3cae16fd067ba9c32a98be22e9b98525", hexFormat.parseHex("3cae16fd067ba9c32a98be22e9b98525")));
        assertThat(ChannelSecret.lookupChannelSecret(hexFormat.parseHex("3cae"))).containsExactly(Map.entry("3cae16fd067ba9c32a98be22e9b98525", hexFormat.parseHex("3cae16fd067ba9c32a98be22e9b98525")));
        assertThat(ChannelSecret.lookupChannelSecret(hexFormat.parseHex("3cae16fd067ba9c32a98be22e9b98525"))).containsExactly(Map.entry("3cae16fd067ba9c32a98be22e9b98525", hexFormat.parseHex("3cae16fd067ba9c32a98be22e9b98525")));
```

yielding this error:

```

java.lang.AssertionError: 
Expecting LinkedList:
  ["3cae16fd067ba9c32a98be22e9b98525"=[60, -82, 22, -3, 6, 123, -87, -61, 42, -104, -66, 34, -23, -71, -123, 37]]
to contain:
  ["3cae16fd067ba9c32a98be22e9b98525"=[60, -82, 22, -3, 6, 123, -87, -61, 42, -104, -66, 34, -23, -71, -123, 37]]
but could not find the following element(s):
  ["3cae16fd067ba9c32a98be22e9b98525"=[60, -82, 22, -3, 6, 123, -87, -61, 42, -104, -66, 34, -23, -71, -123, 37]]
```

**Comment by scordio** [comment]
Hi @msmuenchen, I'm not sure I follow your example.

`LinkedList` has only one type parameter, but you mentioned two: `String` and `byte[]`.

What is `ChannelSecret.lookupChannelSecret()` actually returning?

**Comment by msmuenchen** [comment]
Hi @scordio, the code for reproduction is here: https://github.com/msmuenchen/meshcore-packets-java/blob/main/src/test/java/de/afa_amateurfunk/meshcore_packets/crypto/ChannelSecretTest.java#L70 - a simple `./mvnw test` (after commenting out the `@Disabled` annotation) should show the issue.

`ChannelSecret.lookupChannelSecret` returns a `LinkedList<Map.Entry<String, byte[]>>`. I know, the idiomatic Java way would be to create Yet Another POJO and use `LinkedList<ChannelSecretEntry>`, but I try to avoid too complex object hierarchies for a simple 1:1 map.

**Comment by joel-costigliola** [comment]
I don't think we can address the issue, to do so we need to apply the key comparison semantics when checking the entries are in the same order (can't rely a on key's equals method).
It turns out there is no way to do that, maps don't expose a bipredicate method for the key (or value) comparison semantics.

We need to update the javadoc to explain the limitation.
